Docs: Distinguish examples in rules under Stylistic Issues part 1

pedrottimark opened the pull request:
The first 10 rules under **Stylistic Issues**

Something new in this smaller batch is first attempt at more consistency under **Options**. After some feedback, I will add to the guidelines.

@scriptdaemon If you have time, please critique the changed sentences and bulleted lists under **Options** (and anything else, of course). I applied some of what I am learning through the work you and alberto have done with rule descriptions.

Here are the changes beyond example sentences:
* array-bracket-spacing: delete period in heading; default option first in ul; delete json; delete obsolete sentence about first/last element; move built-in exception from Rule Detail to Options; move Exceptions up to Options
* block-spacing: delete period in heading; delete json; rewrite options
* brace-style: delete As such sentence in Rule Details; delete json; rewrite options
* camelcase: delete json; rewrite options; **please verify** that `"properties": "never"` is an exception option, therefore does not need examples of **incorrect** code
* comma-spacing: move does not apply from end to to Rule Details; delete json; rewrite options; move example with null elements from the end to the options
* comma-style: 2p to ul in Rule Details; delete json; rewrite options; delete empty lines at end of example code
* computed-property-spacing: delete period in heading; default value first; delete json; under Related Rules, delete self-link and replace space-in-brackets with array-bracket-spacing
* consistent-this: rewrite Rule Details; delete json; rewrite options; reorder auxiliary example incorrect/correct
* eol-last: move sentences from intro to Rule Details; rewrite options

scriptdaemon commented on docs/rules/array-bracket-spacing.md:
````diff
@@ -18,26 +18,35 @@ var [x,y] = z;
-"array-bracket-spacing": ["error", "always"]
-```
+* `"never"` allows newlines inside brackets of an array literal, as this is a common pattern
+* `"always"` does not require spaces in an empty array literal `[]`
````
This is one aspect I tried to keep consistent across rule descriptions: Should we use singular or plural here? I.e. "[...] in an empty array literal" or "[...] in empty array literals" ? In the rule descriptions in the README I went with plural where it made sense (I don't have a fundamental reason why I chose plural, I just picked one to keep consistent). Do we want to stick with that for option descriptions as well, or should I change the rule descriptions?

pedrottimark replied:
Yes, I agree with **plural** unless it does not make sense in a particular context.

scriptdaemon commented on docs/rules/array-bracket-spacing.md:
```diff
@@ -18,26 +18,35 @@ var [x,y] = z;
+
+* `"singleValue": true` requires at least one space inside brackets of an array literal which has one element
+* `"objectsInArrays": true` requires at least one space between brackets of an array literal and braces of object literals which are its first or last element
+* `"arraysInArrays": true` requires at least one space between brackets of an array literal and brackets of array literals which are its first or last element
```
"[...] brackets of an array literal [...]" and "[...] brackets of array literals [...]" should both be either singular or plural. Once we decide a preference, I think one should be changes to be consistent with the other.

pedrottimark replied:
Yes, plural. What is a clear fluent way to express the relationship between the array literals?

requires at least one space between brackets of array literals and brackets of array literals which are X first or last elements

* X = the
* X = their
* no word needed at X
* rewrite sentence

scriptdaemon replied:
How about "requires at least one space between brackets of array literals and any adjacent brackets of their containing elements"?

scriptdaemon commented on docs/rules/array-bracket-spacing.md:
```diff
@@ -18,26 +18,35 @@ var [x,y] = z;
+This rule can have an object option for exceptions to the `"never"` option:
+
+* `"singleValue": true` requires at least one space inside brackets of an array literal which has one element
+* `"objectsInArrays": true` requires at least one space between brackets of an array literal and braces of object literals which are its first or last element
```
When we refer to braces, should we always use "curly braces" or simply "braces"? I hadn't really thought of it until now so if we decide that it matters I might need to update my PR as well.

pedrottimark replied:
Good thought! In docs/rules/*.md

* 26 “brace” (sg or pl) are mostly related to blocks (including function bodies)
* 22 “curly brace” (sg or pl) are mostly related to object literals

Let’s use brace/braces consistently and add (also known as …) following the first occurrence in the docs for the 3 rules whose names include curly:

* curly
* object-curly-spacing
* template-curly-spacing

P.S. Just in the past few weeks, I added 9 occurrences of “square bracket” where “bracket” had occurred near “curly brace” so we will get rid of them!

scriptdaemon commented on docs/rules/array-bracket-spacing.md:
```diff
@@ -18,26 +18,35 @@ var [x,y] = z;
+
+This rule can have an object option for exceptions to the `"always"` option:
+
+* `"singleValue": false` disallows spaces inside brackets of an array literal which has one element
```
How about "[...] which has one element" => "[...] that contains one element"?

pedrottimark replied:
:+1:

scriptdaemon commented on docs/rules/brace-style.md:
```diff
@@ -40,27 +40,27 @@ While no style is considered better than the other, most developers agree that h
 ## Rule Details
 
-This rule is aimed at enforcing a particular brace style in JavaScript. As such, it warns whenever it sees a statement or declaration that does not adhere to the one true brace style.
+This rule is aimed at enforcing a particular brace style in JavaScript.
```
Top be consistent with a previous rule in this PR, what if we structured the beginning of this sentence as such "This rule aims to [...]" such that this would become "This rule aims to enforce a particular brace style in JavaScript." and apply it the same way to other rules?

pedrottimark replied:
Yes, “aims to” instead of “is aimed at”

Yesterday, I was thinking that as the rule descriptions become clearer, and when the headings in the rule docs become consistent, this sentence might be redundant. But the repetition might be good: many people seem to scan headings, but not read them expecting to find useful information.

scriptdaemon commented on docs/rules/brace-style.md:
```diff
@@ -40,27 +40,27 @@ While no style is considered better than the other, most developers agree that h
+* `"allman"` enforces Allman style
 
-You can set the style in configuration like this:
+This rule can have an object option for an exception:
```
"[...] can have an [...]" => "[...] has an [...]"

pedrottimark replied:
Yes, “can have” was a feeble attempt to imply the additional option is optional (groan ;-)

Readers will see that from the eslint comments in the example code, true?

scriptdaemon replied:
I actually thought that might have been the case as to why you used that wording, but other options that used "has" were also optional (implied by the existence of "default" values) so I thought it wasn't necessary to break consistency.

I think with the surrounding context (examples, etc.), users should be able to figure out what's required and what isn't.

scriptdaemon commented on docs/rules/comma-spacing.md:
```diff
@@ -13,25 +13,26 @@ var foo = 1 ,bar = 2;
 
 This rule aims to enforce spacing around a comma. As such, it warns whenever it sees a missing or unwanted space in commas of variable declaration, object property, function parameter, sequence and array element.
 
+This rule does not apply in an `ArrayExpression` or `ArrayPattern` to either of the following:
```
"[...] to either of the following" => "[...] for either of the following cases"

scriptdaemon replied:
"to either of the following" => "in either of the following cases"

pedrottimark replied:
Yes, @IanVS helped me with that recently, but I regressed here.

scriptdaemon commented on docs/rules/comma-spacing.md:
```diff
@@ -13,25 +13,26 @@ var foo = 1 ,bar = 2;
+This rule has an object option:
 
-### `{"before": false, "after": true}`
+* `"before": false` (default) disallows space before commas
```
"space" => "spaces" (to be consistent with previous cases where "spaces" is used) ?

pedrottimark replied:
:+1:

scriptdaemon commented on docs/rules/comma-style.md:
```diff
@@ -8,23 +8,33 @@ In case linting is turned off, missing commas in variable declarations lead to l
 ## Rule Details
 
-This rule is aimed at enforcing a particular comma style in JavaScript. As such, it warns whenever it sees a variable declaration, object property and array element that does not adhere to a particular comma style. It doesn't support cases where there are line breaks before and after comma (lone commas) with in declarations, properties and elements. It also avoids single line declaration cases.
+This rule is aimed at enforcing a particular comma style in JavaScript. As such, it warns whenever it sees a variable declaration, object property and array element that does not adhere to a particular comma style.
```
"As such, it warns whenever a variable declaration, object property, or array element does not adhere to a particular comma style." (or the respective changes if we decide to go with plural.)

pedrottimark replied:
How about one sentence?

This rule aims to enforce consistent comma style in variable declarations, object literals, and array literals.

By the way, a team member suggested “report” as a verb instead of “warn” to avoid confusion with error versus warning, but I think we will solve that inconsistency as we simplify grammar and move to verbs like allow, disallow, enforce, require.

scriptdaemon replied:
Shorter is sweeter, in my option. :+1: 

I'm glad you mentioned that, as the use of "warn" in that context has bugged me in the past because it does confuse what should be "report" with the actual "warn" severity. As we simplify the docs to be more concise, it might not be a problem anymore, but I prefer the use of "report" over "warn" if there ends up being a sentence that still uses it.

scriptdaemon commented:
@pedrottimark Looks good overall. My suggestions are only minor. Mostly, I think tense, singular vs plural, etc. should remain consistent across the docs (and should also make sense in their respective contexts).

pedrottimark commented:
@scriptdaemon Thank you very much! That is a super example of critique.

pedrottimark commented:
Updated after first critique and also changed:
* Under Rule Details, replaced “This rule aims to …” or “This rule is aimed at …” with “This enforces …” from the new rule descriptions, and sometimes added more details from a previous second sentence.
* In the bulleted items under options, replaced “enforces” with “requires” to get feedback from reviewers if that is any better contrast with “disallows”

scriptdaemon commented on docs/rules/array-bracket-spacing.md:
```diff
@@ -1,9 +1,9 @@
-A number of style guides require or disallow spaces between array brackets. This rule
-applies to both array literals and destructuring assignment (EcmaScript 6) using arrays.
+A number of style guides require or disallow spaces between array brackets and other tokens. This rule
+applies to both array literals and destructuring assignment (ECMAScript 6) using arrays.
```
"destructuring assignment" => "destructuring assignments"

pedrottimark replied:
:+1:

scriptdaemon commented:
One last inline suggestion, but LGTM.

pedrottimark commented:
In array-bracket-spacing, appended little inline code examples to descriptions of objectsInArrays and arraysInArrays options.
